packages/utils/src/lock.ts

```diff
@@ -0,0 +1,54 @@
+import { EventEmitter } from 'events'
+
+enum LockEvent {
+  Unlock = 'unlock',
+}
+
+// Lock which can be used to ensure mutual exclusion in concurrent code.
+export class Lock {
```

Comment:
surprised there isn't an obvious library we could use for this?

Reply:
We could pull in a library, but I opted not to when I realized the [most popular implementation](https://www.npmjs.com/package/async-lock) has a big disclaimer about how the maintainer is not active 😕 . I figure it's not worth the incremental attack surface increase and we are better off having a simple implementation locally.

Reply:
I thought the same and was trying to google search for one 🤔